crates/walrus-service/src/node.rs

```diff
@@ -2717,6 +2902,155 @@ impl StorageNodeInner {
+                walrus.blob_id = %blob_id,
+                "skipping recovery deferral; blob already stored on all shards"
+            );
+            return Err(SetRecoveryDeferralError::AlreadyStored);
+        }
+
+        let until = std::time::Instant::now() + defer_duration;
+        let mut map = self.recovery_deferrals.write().await;
+        if map.contains_key(&blob_id) {
+            tracing::trace!(
+                walrus.blob_id = %blob_id,
+                "recovery deferral already exists for blob"
+            );
+            return Err(SetRecoveryDeferralError::AlreadyExists);
+        }
+
+        map.insert(
+            blob_id,
+            (
+                until,
+                std::sync::Arc::new(tokio_util::sync::CancellationToken::new()),
+            ),
+        );
+        let current_size = map.len();
+        self.update_recovery_deferral_size_metric(current_size);
+        drop(map);
+
+        self.recovery_deferral_notify.notify_one();
+
+        Ok(())
+    }
+
+    /// Removes any active recovery deferral for the specified blob.
+    pub async fn clear_recovery_deferral(&self, blob_id: &BlobId) {
+        let mut map = self.recovery_deferrals.write().await;
+        if let Some((_, token)) = map.remove(blob_id) {
+            token.cancel();
+        }
+        self.update_recovery_deferral_size_metric(map.len());
+        drop(map);
+        self.recovery_deferral_notify.notify_one();
+    }
+
+    fn update_recovery_deferral_size_metric(&self, size: usize) {
+        // Avoid potential wrap by saturating at i64::MAX on extremely large sizes.
+        let size_i64 = i64::try_from(size).unwrap_or(i64::MAX);
+        self.metrics.recovery_deferrals_active.set(size_i64);
+    }
+
+    fn start_recovery_deferral_cleanup_task(self: &Arc<Self>) {
+        let this = Arc::clone(self);
+        tokio::spawn(async move {
+            this.recovery_deferral_cleanup_task().await;
+        });
+    }
+
+    async fn recovery_deferral_cleanup_task(self: Arc<Self>) {
+        loop {
+            if self.recovery_deferral_cleanup_token.is_cancelled() {
+                break;
+            }
+
+            self.remove_expired_recovery_deferrals().await;
+
+            let sleep_duration = match self.next_recovery_deferral_deadline().await {
+                Some(deadline) => deadline.saturating_duration_since(std::time::Instant::now()),
+                None => Duration::from_secs(60),
+            };
+
+            // Any newly scheduled deferral notifies the waiters below, forcing us to recompute the
+            // minimum deadline so we never oversleep a shorter timeout.
+            tokio::select! {
+                _ = self.recovery_deferral_cleanup_token.cancelled() => break,
+                _ = self.recovery_deferral_notify.notified() => continue,
+                _ = tokio::time::sleep(sleep_duration) => {}
+            }
+        }
+
+        self.cancel_all_recovery_deferrals().await;
```

Comment:
Suggestion: also disable recovery deferral in the node. Existing this loop won't cause node to stop, and future blob sync may still use deferral to start blob sync, and will never be started.

Reply:
Yes, but i think it is better to call `cancel` on the token since we hold a reference to it.